hnn_core/network.py

```diff
@@ -219,47 +295,430 @@ def __repr__(self):
+            space_constant, synaptic_delays, cell_specific=cell_specific)
+
+        # Must remember to update the GID ranges based on pos_dict!
+        self.pos_dict[name] = [self.pos_dict['origin'] for
+                               dg in src_gid_ran]
+
+        # NB _update_gid_ranges checks external_drives[name] for drives!
+        self.external_drives[name] = drive
         # Every time pos_dict is updated, gid_ranges must be updated too
         self._update_gid_ranges()
 
-        # Create the feed dynamics (event_times)
-        self._instantiate_feeds(n_trials=self.params['N_trials'])
+    def _create_drive_conns(self, target_populations, weights_by_receptor,
+                            location, space_constant, synaptic_delays,
+                            cell_specific=True):
+        """Create parameter dictionary defining how drive connects to network
+
+        Parameters
+        ----------
+        target_populations : list
+            Cell names/types to attach the drive to
+        weights_by_receptor : dict (keys: 'ampa' and 'nmda')
+            Synaptic weights for each receptor type
+        location : str
+            Target location of synapses ('distal' or 'proximal')
+        space_constant : float
+            Describes lateral dispersion (from column origin) of synaptic
+            weights and delays within the simulated column
+        synaptic_delays : dict
+            Synaptic delay at the column origin, dispersed laterally as a
+            function of the space_constant
+        cell_specific : bool
+            Whether each cell has unique connection parameters (default: True)
+            or all cells have common connections to a global (single) drive.
 
-    def _instantiate_feeds(self, n_trials=1):
-        """Creates event_time vectors for all feeds and all trials
+        Returns
+        -------
+        drive_conn_by_cell : dict
+            Keys are target_populations, each item is itself a dict with keys
+            relevant to how the drive connects to the rest of the network.
+        src_gid_ran : range
+            For convenience, return back the range of GIDs associated with
+            all the driving units (they become _ArtificialCell's later)
+
+        Note
+        ----
+        drive_conn : dict
+            'target_gids': range of target cell GIDs
+            'target_type': target cell type (e.g. 'L2_basket')
+            'src_gids': range of source (artificial) cell GIDs
+            'location': 'distal' or 'proximal'
+            'ampa' and 'nmda': dict
+                'A_weight': synaptic weight
+                'A_delay':  synaptic delay (used with space constant)
+                'lamtha': space constant
+        """
+        if isinstance(synaptic_delays, dict):
+            for receptor in ['ampa', 'nmda']:
+                if not (
+                    set(list(weights_by_receptor[receptor].keys())).issubset(
+                        set(list(synaptic_delays.keys())))):
```

Comment:
it's a mouthful to read. Shouldn't they have the same keys? Just do:

```py
if not weights_by_receptor[receptor].keys() == synaptic_delays.keys():
```

in any case the `list()` seems superfluous above

Reply:
Turns out `dict_keys` type cannot be cast to a `set`. I use sets because I actually don't expect the lists of keys to be equal. I'll run a pass over those checks though to see if I can simplify...

Reply:
Turns out it can (be cast)... Must have used the wrong bag of mushrooms for that soup...

Reply:
Also: yes, I want them to be equal, not just subsets!

Reply:
Equality test still not OK though: there's only one `synaptic_delays` dict, applied to both AMPA and NMDA weights (which may have keys missing, hence `issubset`)

Future PR: maybe weights and delays should be assigned as a tuple?

```python
ampa_synapses = {'L2_pyramidal': (0.01, 1.0), ...}  # (uS, ms)
```